src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs

```diff
@@ -88,72 +88,104 @@ public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContex
     // This is the heart of the activity and actually performs the work to be done.
     public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
     {
-        var isValid = false;
-        IFormCollection form = null;
-        string email = null;
-        if (_httpContextAccessor.HttpContext != null)
+        var email = GetPropertyFromFormOrContext(workflowContext, "Email");
+        if (string.IsNullOrWhiteSpace(email))
         {
-            form = _httpContextAccessor.HttpContext.Request.Form;
-            email = form["Email"];
-            isValid = !string.IsNullOrWhiteSpace(email);
+            return Outcomes("Done", "Invalid");
         }
-        var outcome = isValid ? "Valid" : "Invalid";
 
-        if (isValid)
+        var userName = GetPropertyFromFormOrContext(workflowContext, "UserName") ?? email.Replace('@', '+');
+        var user = await CreateUserAsync(userName, email);
+
+        if (user == null)
         {
-            var userName = form["UserName"];
-            if (string.IsNullOrWhiteSpace(userName))
-            {
-                userName = email.Replace('@', '+');
-            }
+            return Outcomes("Done", "Invalid");
+        }
+
+        if (SendConfirmationEmail && !await SendConfirmationEmailAsync(user, workflowContext, email))
+        {
+            return Outcomes("Done", "Invalid");
+        }
+
+        return Outcomes("Done", "Valid");
+    }
+
+    private string GetPropertyFromFormOrContext(WorkflowExecutionContext context, string key)
+    {
+        var form = _httpContextAccessor.HttpContext?.Request.Form;
+        var value = form?[key];
+
+        if (!string.IsNullOrEmpty(value))
+        {
+            return value;
+        }
+
+        if (context.Properties.TryGetValue(key, out var objValue) && objValue is string strValue)
```

Comment:
Since this is after the `Form` access, if one sets this from the workflow context, a malicious user can overwrite it from the request. So, better would be to check for this first, and fall back to the form data.

Reply:
Done :-)